source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,154 @@
+
+Until https://bugzilla.redhat.com/1855367 is implemented in libvirt, VMs with TPM devices cannot have snapshots with memory.  The problem is that we need swtpm data snapshot consistent with the VM snapshot, which is not possible to get without libvirt assistance.
+
+Since the emulated TPM is a software rather than a hardware device, its data updates can get lost, e.g. when a host suddenly crashes.  (TODO: Clarify according to the selected implementation.)
+
+There is apparently no guarantee that new swtpm versions can still read its old data.  That should be watched in testing.
+
+## User Experience
+
+Ideally, a TPM device would be added to each VM automatically and would be automatically available to the guest OS.  However, TPM support is a new feature and we cannot guarantee that everything is correctly supported in the underlying platform (for example, snapshots with memory are currently not working as expected).  For that reason, it's better to not add TPM devices to VMs by default for now and make a cluster option to enable them for the given cluster, similar to memory ballooning.
+
+TODO: Description of the UI actions once they are implemented.
+
+## Testing
+
+There are two areas to test: TPM presence and data persistence.
+
+TPM devices should be present in VMs running in clusters of versions >= 4.4 (TODO: re-check the version) when the corresponding cluster option is checked.  A TPM device is present as ``<tpm>`` device in the domain XML and as a corresponding device in the guest OS (``/dev/tpm0`` on Linux systems).
+
+Newly created VMs, or VMs that didn't have a TPM device previously, get uninitialized TPM data, there is nothing to test about it.  Once TPM data is modified by the guest OS, the modified data must be present after different operations: VM stop & start, reboot, migration, export/import, cloning, start from snapshot, etc.  TPM data updates followed by common failure scenarios (disconnected host, crashed VM, etc.) can be tested in situations where it's reasonable to assume the data updates should be available.  The data can be checked in the guest OS (most reliably) or alternatively comparing swtpm data with expected contents.
+
+It should be also tested that VMs with TPM data created on hosts running older systems are unchanged when the VM is started on a host running the newest system.  This must be tested in the guest OS.
+
+## Changes to Engine and Vdsm
+
+There are no special changes needed in order to add a TPM device as such, it's just another kind of device added to Engine.
+
+### Data persistence
+
+A very different thing is ensuring TPM data persistence.  TPM data is managed using an external tool, swtpm.  Once the VM domain XML is deleted on the host, the TPM data is deleted as well by libvirt and it must be restored before the VM is run again, whether on this or another host (live host migrations are handled transparently and don't require any special actions on the oVirt side).
```

Comment:
How live migration is handled? how it does not require anything on ovirt side?

Reply:
Live migration is handled transparently by libvirt. The only trouble with it could be if swtpm was arranged to write to the same shared storage volume directly on both the source and the destination.